src/core/kdtree.hpp

```diff
@@ -0,0 +1,457 @@
+	{
+		this->unbalanced += amount;
+	}
+
+	bool IsUnbalanced()
+	{
+		size_t count = this->Count();
+		if (count < 8) return false;
+		return this->unbalanced > this->Count() / 4;
+	}
+
+	void CheckInvariant(size_t node_idx, int level, CoordT min_x, CoordT max_x, CoordT min_y, CoordT max_y)
+	{
+		if (node_idx == INVALID_NODE) return;
+
+		const auto &n = this->nodes[node_idx];
+		CoordT cx = this->xyfunc(n.element, 0);
+		CoordT cy = this->xyfunc(n.element, 1);
+
+		assert(cx >= min_x);
+		assert(cx < max_x);
+		assert(cy >= min_y);
+		assert(cy < max_y);
+
+		if (level % 2 == 0) {
+			// split in dimension 0 = x
+			CheckInvariant(n.left,  level + 1, min_x, cx, min_y, max_y);
+			CheckInvariant(n.right, level + 1, cx, max_x, min_y, max_y);
+		} else {
+			// split in dimension 1 = y
+			CheckInvariant(n.left,  level + 1, min_x, max_x, min_y, cy);
+			CheckInvariant(n.right, level + 1, min_x, max_x, cy, max_y);
+		}
+	}
+
+	void CheckInvariant()
+	{
+#ifdef KDTREE_DEBUG
+		CheckInvariant(this->root, 0, std::numeric_limits<CoordT>::min(), std::numeric_limits<CoordT>::max(), std::numeric_limits<CoordT>::min(), std::numeric_limits<CoordT>::max());
+#endif
+	}
+
+public:
+	/** Construct a new Kdtree with the given xyfunc */
+	Kdtree(TxyFunc xyfunc) : root(INVALID_NODE), xyfunc(xyfunc), unbalanced(0) { }
+
+	/**
+	 * Clear and rebuild the tree from a new sequence of elements,
+	 * @tparam It    Iterator type for element sequence.
+	 * @param  begin First element in sequence.
+	 * @param  end   One past last element in sequence.
+	 */
+	template <typename It>
+	void Build(It begin, It end)
+	{
+		this->nodes.clear();
+		this->free_list.clear();
+		this->unbalanced = 0;
+		if (begin == end) return;
+		this->nodes.reserve(end - begin);
+
+		this->root = this->BuildSubtree(begin, end, 0);
+		CheckInvariant();
+	}
+
+	/**
+	 * Reconstruct the tree with the same elements, letting it be fully balanced.
+	 */
+	void Rebuild()
+	{
+		this->Rebuild(NULL, NULL);
+	}
+
+	/**
+	 * Insert a single element in the tree.
+	 * Repeatedly inserting single elements may cause the tree to become unbalanced.
+	 * Undefined behaviour if the element already exists in the tree.
+	 */
+	void Insert(const T &element)
+	{
+		if (this->Count() == 0) {
+			this->root = this->AddNode(element);
+		} else {
+			if (!this->IsUnbalanced() || !this->Rebuild(&element, NULL)) {
+				this->InsertRecursive(element, this->root, 0);
+				this->IncrementUnbalanced();
+			}
+			CheckInvariant();
+		}
+	}
+
+	/**
+	 * Remove a single element from the tree, if it exists.
+	 * Since elements are stored in interior nodes as well as leaf nodes, removing one may
+	 * require a larger sub-tree to be re-built. Because of this, worst case run time is
+	 * as bad as a full tree rebuild.
+	 */
+	void Remove(const T &element)
+	{
+		size_t count = this->Count();
+		if (count == 0) return;
+		if (!this->IsUnbalanced() || !this->Rebuild(NULL, &element)) {
+			this->RemoveRecursive(element, this->root, 0);
```

Comment:
It seems like `RemoveRecursive` can modify `this->root`. Is it possible to emphasize this behavior? Maybe by changing he return value to be the new node index, and have the original bool value given as a return parameter, or compute it (whether the element to remove was found) from the returning node index (whether it was changed or not).

Reply:
I'll add a few words about that.